actors/miner/src/lib.rs

```diff
@@ -2995,6 +2996,297 @@ impl Actor {
+                                    params.from_deadline
+                                ));
+                            }
+                        }
+
+                        // Load sector infos for proof, substituting a known-good sector for known-faulty sectors.
+                        // Note: this is slightly sub-optimal, loading info for the recovering sectors again after they were already
+                        // loaded above.
+                        let sector_infos = sectors
+                            .load_for_proof(
+                                &BitField::union(&all_sectors),
+                                &BitField::union(&all_ignored),
+                            )
+                            .map_err(|e| {
+                                e.downcast_default(
+                                    ExitCode::USR_ILLEGAL_STATE,
+                                    "failed to load sectors for post verification",
+                                )
+                            })?;
+                        if !verify_windowed_post(
+                            rt,
+                            current_deadline.challenge,
+                            &sector_infos,
+                            window_proof.proofs.clone(),
+                        )
+                        .map_err(|e| e.wrap("window post failed"))?
+                        {
+                            return Err(actor_error!(
+                                illegal_argument,
+                                "invalid post was submitted"
+                            )
+                            .into());
+                        }
+                        Ok(())
+                    })
+                    .map_err(|e| {
+                        e.downcast_default(ExitCode::USR_ILLEGAL_STATE, "while removing partitions")
+                    })?;
+            }
+
+            if !deadline_available_for_compaction(
+                policy,
+                current_deadline.period_start,
+                params.to_deadline,
+                rt.curr_epoch(),
+            ) {
+                return Err(actor_error!(
+                    forbidden,
+                    "cannot move to deadline {} during its challenge window, \
+                        or the prior challenge window,
+                        or before {} epochs have passed since its last challenge window ended",
+                    params.to_deadline,
+                    policy.wpost_dispute_window
+                ));
+            }
+
+            if !deadline_available_for_move(
+                policy,
+                params.from_deadline,
+                params.to_deadline,
+                current_deadline.index,
+            ) {
+                return Err(actor_error!(
+                    forbidden,
+                    "cannot move to a deadline which is further from current deadline"
+                ));
+            }
```

Comment:
Could we move this before the transaction handling? 
I would think it is better if it can fail faster.

Reply:
Oops, seems not , because `current_deadline` depends on `state`, which is only available in the transaction.